apollo-router/src/spec/query/change.rs

```diff
@@ -132,158 +214,268 @@ impl<'a> QueryHashVisitor<'a> {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
 
                 self.hash_join_type(&o.name, &o.directives)?;
+                "^OBJECT_END".hash(self);
+
+                "^IMPLEMENTED_INTERFACES_LIST".hash(self);
+                for interface in &o.implements_interfaces {
+                    self.hash_type_by_name(&interface.name)?;
+                }
+                "^IMPLEMENTED_INTERFACES_LIST_END".hash(self);
             }
             ExtendedType::Interface(i) => {
+                "^INTERFACE".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &i.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
+
                 self.hash_join_type(&i.name, &i.directives)?;
+
+                "^IMPLEMENTED_INTERFACES_LIST".hash(self);
+                for implementor in &i.implements_interfaces {
+                    self.hash_type_by_name(&implementor.name)?;
+                }
+                "^IMPLEMENTED_INTERFACES_LIST_END".hash(self);
+
+                if let Some(implementers) = self.schema().implementers_map().get(&i.name) {
+                    "^IMPLEMENTER_OBJECT_LIST".hash(self);
+
+                    for object in &implementers.objects {
+                        self.hash_type_by_name(object)?;
+                    }
+                    "^IMPLEMENTER_OBJECT_LIST_END".hash(self);
+
+                    "^IMPLEMENTER_INTERFACE_LIST".hash(self);
+                    for interface in &implementers.interfaces {
+                        self.hash_type_by_name(interface)?;
+                    }
+                    "^IMPLEMENTER_INTERFACE_LIST_END".hash(self);
+                }
+
+                "^INTERFACE_END".hash(self);
             }
             ExtendedType::Union(u) => {
+                "^UNION".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &u.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIS_END".hash(self);
 
+                "^MEMBER_LIST".hash(self);
                 for member in &u.members {
                     self.hash_type_by_name(member.as_str())?;
                 }
+                "^MEMBER_LIST_END".hash(self);
             }
             ExtendedType::Enum(e) => {
+                "^ENUM".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &e.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
 
+                "^ENUM_VALUE_LIST".hash(self);
                 for (value, def) in &e.values {
+                    "^VALUE".hash(self);
+
                     value.hash(self);
                     for directive in &def.directives {
```

Comment:
All other directives lists are wrapped in `^DIRECTIVE_LIST`.
No exploitable, just consistency.
Maybe it worth to create a function to hash directive lists?